content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,249 @@
+
+{{% tweet 988527557300568065 %}}
+
+It’s been a while since I started reading more papers, following awesome initiatives like [Papers We Love](https://paperswelove.org/) and [The Morning Paper](https://blog.acolyer.org/) — I always wanted to experiment with publishing notes. This will be the first attempt.
+
+{{% center %}} … {{% /center %}}
+
+[“Lessons from Building Static Analysis Tools at Google](https://cacm.acm.org/magazines/2018/4/226371-lessons-from-building-static-analysis-tools-at-google/fulltext)” by Caitlin Sadowski, Edward Aftandilian, [Alex Eagle](undefined), Liam Miller-Cushon, Ciera Jaspan presents 2 stories: history of failed attempts of integrating [FindBugs](https://github.com/findbugsproject/findbugs), a static analysis tool for Java at [Google](http://twitter.com/Google), and lessons learned from success story of incorporating extensible analysis framework, a [Tricorder project](https://research.google.com/pubs/pub43322.html), to development workflow at Google.
+
+## Source code analysis recap
+
+Before digging deeper into the paper, a bit of the context on what kind of analysis in general is applicable to programs. There are two main type of program analysis:
+
+* *Static code analysis*, looks only at the source code, without running the program
+* *Dynamic code analysis*, when (potentially customized) program is executed and results are analyzed
+
+This paper talks only about *static code analysis*, which is traditionally further subdivided into two main flavors:
+
+* *intra-procedural,* when analysis is localized inside a small part of the program like a function/class/file
+* *inter-procedural,* that uses broader, whole-program artefacts as input, and conduct more sophisticated analysis (call graph, formal logic, context-sensitive analysis, etc)
+
+According to the paper, Google only runs simpler or *intra-procedural* type of analysis — the only one feasible to conduct at scale of 2 bln LoC.
+
+Interesting enough, this is somehow different from Facebook’s project [Infer](http://fbinfer.com/) approach, which developed a way scale a particular type of [more complex](https://code.facebook.com/posts/1537144479682247/finding-inter-procedural-bugs-at-scale-with-infer-static-analyzer/), *inter-procedural* [separation logic-based](https://en.wikipedia.org/wiki/Separation_logic) compositional analysis, that paper authors are well aware of.
+
+[**Infer static analysis tool**](http://fbinfer.com/)
+TK does hugo have nice big links?
+
+The reasoning behind why Google is not interested in more complex analysis at this point are explicitly listed in the paper:
+
+* *Large investment.* Although theoretically better and more complex analysis exist, it will require non-trivial engineering effort to scale
+
+* *High upfront cost.* It’s very hard to get the data, even to just estimate the cost-benefit ratio for such improvements, to justify research/implementation efforts
+
+* *Work still would be needed to reduce false-positive rates*
```

Comment:
reduce the number of false positives

Reply:
Although agree, but would prefer to keep as this is the term they consistently use though the paper

Reply:
not en expert, but would prefer keep it as is with dash, as that is exactly the way it's written in the original paper

Reply:
The original paper can use whatever they want, after all, they are not stats or ML experts. I read inconsistent names in papers every week but it is not the reason to spread them. I can assure you that nobody is going to be confused if you drop the dash from "false-positives" whereas the dashed variant annoys the nazis.

But anyway @campoy or @vcoisne know better.